

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>treasmo.core &#8212; TREASMO v1.1.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=365ca57ee442770a23c6" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=365ca57ee442770a23c6" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=365ca57ee442770a23c6" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=365ca57ee442770a23c6"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/treasmo/core';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">TREASMO v1.1.0 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">



<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <h1>Source code for treasmo.core</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="kn">import</span> <span class="nn">time</span>

<span class="kn">from</span> <span class="nn">treasmo.moran_vec</span> <span class="kn">import</span> <span class="n">Moran</span>
<span class="kn">import</span> <span class="nn">treasmo.lee_vec</span> <span class="k">as</span> <span class="nn">lee_vec</span>
<span class="kn">from</span> <span class="nn">libpysal.weights</span> <span class="kn">import</span> <span class="n">W</span>

<span class="kn">import</span> <span class="nn">scipy</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="c1">#from statsmodels.stats.multitest import multipletests</span>
<span class="kn">from</span> <span class="nn">statsmodels.stats.multitest</span> <span class="kn">import</span> <span class="n">fdrcorrection</span>

<span class="n">__author__</span> <span class="o">=</span> <span class="s2">&quot;Chaozhong Liu&quot;</span>
<span class="n">__email__</span> <span class="o">=</span> <span class="s2">&quot;czliubioinfo@gmail.com&quot;</span>



<div class="viewcode-block" id="Morans_I"><a class="viewcode-back" href="../../api.html#treasmo.core.Morans_I">[docs]</a><span class="k">def</span> <span class="nf">Morans_I</span><span class="p">(</span><span class="n">mudata</span><span class="p">,</span> <span class="n">mods</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">,</span><span class="s1">&#39;atac&#39;</span><span class="p">],</span> <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_RAM</span><span class="o">=</span><span class="mi">16</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    Function to calculate Moran&#39;s I for all the features in multiome data</span>

<span class="sd">    Arguments</span>
<span class="sd">    --------------</span>

<span class="sd">        mudata: MuData</span>
<span class="sd">            single-cell multi-omics data saved as MuData object</span>

<span class="sd">        mods: List[str, str]</span>
<span class="sd">            scRNA-seq and scATAC-seq modality name in MuData object</span>

<span class="sd">        seed: int</span>
<span class="sd">            random seed to make the results reproducible</span>

<span class="sd">        max_RAM: int</span>
<span class="sd">            maximum limitation of memory usage (Gb)</span>


<span class="sd">    Returns</span>
<span class="sd">    --------------</span>
<span class="sd">        MuData</span>
<span class="sd">            added with var[&#39;Morans.I&#39;]</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Check if mods exist in MuData</span>
    <span class="k">if</span> <span class="nb">sum</span><span class="p">([</span><span class="n">i</span> <span class="ow">in</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">mods</span><span class="p">])</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">mods</span><span class="p">):</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Not all mods provided are in MuData! Running stopped.&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="c1"># Start</span>
    <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="c1"># Construct connection matrix</span>
    <span class="n">cong_mtx</span> <span class="o">=</span> <span class="n">mudata</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s1">&#39;connectivities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
    <span class="c1">#cong_mtx = cong_mtx/cong_mtx.sum(axis=1)[:,np.newaxis]</span>
    <span class="n">neighbors</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">weights</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cong_mtx</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">neighbors</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">cong_mtx</span><span class="p">[</span><span class="n">i</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">weights</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">cong_mtx</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">cong_mtx</span><span class="p">[</span><span class="n">i</span><span class="p">])[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">w</span> <span class="o">=</span> <span class="n">W</span><span class="p">(</span><span class="n">neighbors</span><span class="p">,</span> <span class="n">weights</span><span class="p">)</span>

    <span class="n">mi</span> <span class="o">=</span> <span class="n">Moran</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,</span> <span class="n">w</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">mudata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;Morans.I&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">NaN</span>

    <span class="k">for</span> <span class="n">mod</span> <span class="ow">in</span> <span class="n">mods</span><span class="p">:</span>
        <span class="n">mi</span><span class="o">.</span><span class="n">calc_i</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span> <span class="n">max_RAM</span><span class="o">=</span><span class="n">max_RAM</span><span class="p">)</span>
        <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;Morans.I&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mi</span><span class="o">.</span><span class="n">I</span>

    <span class="n">mudata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;Morans.I&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;Morans.I&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span> <span class="k">for</span> <span class="n">mod</span> <span class="ow">in</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>
    
    <span class="c1">#print(mi.I)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Finished calculating Moran&#39;s I. </span><span class="si">%.3f</span><span class="s2">s past&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">mudata</span></div>




<span class="k">def</span> <span class="nf">_check_array</span><span class="p">(</span><span class="n">mtx</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">sp</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">issparse</span><span class="p">(</span><span class="n">mtx</span><span class="p">):</span> <span class="c1">#isinstance(mtx, sp.spmatrix):</span>
        <span class="k">return</span> <span class="n">mtx</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mtx</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">mtx</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Omics Data should be either numpy array or scipy sparse matrix.&quot;</span><span class="p">)</span>




<span class="k">def</span> <span class="nf">_prepare_mtx</span><span class="p">(</span><span class="n">anndat</span><span class="p">,</span> <span class="n">features</span><span class="p">):</span>

    <span class="n">index_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;index&#39;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">anndat</span><span class="o">.</span><span class="n">n_vars</span><span class="p">)})</span>
    <span class="n">index_df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">anndat</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span>

    <span class="n">feature_index</span> <span class="o">=</span> <span class="n">index_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">features</span><span class="p">,</span><span class="s1">&#39;index&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">_check_array</span><span class="p">(</span><span class="n">anndat</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span><span class="n">feature_index</span><span class="p">])</span>




<div class="viewcode-block" id="Global_L"><a class="viewcode-back" href="../../api.html#treasmo.core.Global_L">[docs]</a><span class="k">def</span> <span class="nf">Global_L</span><span class="p">(</span><span class="n">mudata</span><span class="p">,</span> <span class="n">pairsDf</span><span class="p">,</span> <span class="n">mods</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">,</span><span class="s1">&#39;atac&#39;</span><span class="p">],</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">percent</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_RAM</span><span class="o">=</span><span class="mi">16</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    Function to calculate the global L index (mean of correlation strength index) for all the pairs in multiome data</span>

<span class="sd">    Arguments</span>
<span class="sd">    --------------</span>

<span class="sd">        mudata: MuData</span>
<span class="sd">            single-cell multi-omics data saved as MuData object</span>

<span class="sd">        mods: List[str, str]</span>
<span class="sd">            scRNA-seq and scATAC-seq modality name in MuData object</span>

<span class="sd">        pairsDf: pandas.DataFrame</span>
<span class="sd">            gene-peak pair DataFrame containing pairs to be calculated</span>
<span class="sd">            self-prepared or called from treasmo.tl.peaks_within_distance / TFBS_match</span>

<span class="sd">        permutations: int</span>
<span class="sd">            Number of permutations for significance test.</span>
<span class="sd">            Default is 0, meaning no significance test</span>
<span class="sd">            999 is a good choice for most cases, but it might take a long time (hours) to finish depending on the number of pairs</span>
<span class="sd">        </span>
<span class="sd">        percent: float</span>
<span class="sd">            percentage of cells to shuffle during permutation.</span>
<span class="sd">            For most of the time, default 0.1 is already a good choice.</span>
<span class="sd">        </span>
<span class="sd">        seed: int</span>
<span class="sd">            random seed to make the results reproducible</span>

<span class="sd">        max_RAM: int</span>
<span class="sd">            maximum limitation of memory usage(Gb)</span>


<span class="sd">    Returns</span>
<span class="sd">    --------------</span>
<span class="sd">        DataFrame</span>
<span class="sd">            pairsDf added with extra columns:</span>
<span class="sd">                Global L results</span>
<span class="sd">                QC metrics (feature sparsity)</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Calculating KNN graph-based global correlation...&quot;</span><span class="p">)</span>
    <span class="n">cong_mtx</span> <span class="o">=</span> <span class="n">mudata</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s1">&#39;connectivities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
    <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="n">cong_mtx</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
    <span class="n">eSP</span> <span class="o">=</span> <span class="n">lee_vec</span><span class="o">.</span><span class="n">Spatial_Pearson</span><span class="p">(</span><span class="n">cong_mtx</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="n">permutations</span><span class="p">)</span>

    <span class="n">genes</span> <span class="o">=</span> <span class="n">pairsDf</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span>
    <span class="n">peaks</span> <span class="o">=</span> <span class="n">pairsDf</span><span class="p">[</span><span class="s1">&#39;Peak&#39;</span><span class="p">]</span>

    <span class="n">gene_X</span> <span class="o">=</span> <span class="n">_prepare_mtx</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">genes</span><span class="p">)</span>
    <span class="n">peak_X</span> <span class="o">=</span> <span class="n">_prepare_mtx</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span> <span class="n">peaks</span><span class="p">)</span>

    <span class="n">eSP</span> <span class="o">=</span> <span class="n">eSP</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">gene_X</span><span class="p">,</span> <span class="n">peak_X</span><span class="p">,</span> <span class="n">percent</span><span class="o">=</span><span class="n">percent</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span> <span class="n">max_RAM</span><span class="o">=</span><span class="n">max_RAM</span><span class="p">)</span>

    <span class="n">pairsDf</span><span class="p">[</span><span class="s1">&#39;L&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">eSP</span><span class="o">.</span><span class="n">association_</span>
    <span class="n">pairsDf</span><span class="p">[</span><span class="s1">&#39;L.p_value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">eSP</span><span class="o">.</span><span class="n">significance_</span> <span class="k">if</span> <span class="n">permutations</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">NaN</span>

    <span class="k">if</span> <span class="n">permutations</span><span class="p">:</span>
        <span class="n">_</span><span class="p">,</span><span class="n">pairsDf</span><span class="p">[</span><span class="s1">&#39;L.FDR&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fdrcorrection</span><span class="p">(</span><span class="n">peaks_nearby</span><span class="p">[</span><span class="s1">&#39;L.p_value&#39;</span><span class="p">],</span>
                                           <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;indep&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">pairsDf</span><span class="p">[</span><span class="s1">&#39;L.FDR&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">NaN</span>

    <span class="n">pairsDf</span><span class="p">[</span><span class="s1">&#39;gene.pct&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">genes</span><span class="p">,</span><span class="s1">&#39;Frac.all&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">pairsDf</span><span class="p">[</span><span class="s1">&#39;peak.pct&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">peaks</span><span class="p">,</span><span class="s1">&#39;Frac.all&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Finished calculating correlation. </span><span class="si">%.3f</span><span class="s2">s past&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">pairsDf</span></div>




<span class="k">def</span> <span class="nf">_calculate_LL</span><span class="p">(</span><span class="n">mudata</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">peaks</span><span class="p">,</span>
                  <span class="n">mods</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">,</span><span class="s1">&#39;atac&#39;</span><span class="p">],</span>
                  <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_RAM</span><span class="o">=</span><span class="mi">16</span><span class="p">):</span>
    <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>

    <span class="n">cong_mtx</span> <span class="o">=</span> <span class="n">mudata</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s1">&#39;connectivities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
    <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="n">cong_mtx</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>

    <span class="n">gene_X</span> <span class="o">=</span> <span class="n">_prepare_mtx</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">genes</span><span class="p">)</span>
    <span class="n">peak_X</span> <span class="o">=</span> <span class="n">_prepare_mtx</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span> <span class="n">peaks</span><span class="p">)</span>

    <span class="n">eSP2</span> <span class="o">=</span> <span class="n">lee_vec</span><span class="o">.</span><span class="n">Spatial_Pearson_Local</span><span class="p">(</span><span class="n">cong_mtx</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">eSP2</span> <span class="o">=</span> <span class="n">eSP2</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">gene_X</span><span class="p">,</span><span class="n">peak_X</span><span class="p">,</span><span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span> <span class="n">max_RAM</span><span class="o">=</span><span class="n">max_RAM</span><span class="p">)</span>

<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    if permutations:</span>
<span class="sd">        local_p_df = pd.DataFrame(eSP2.significance_)</span>
<span class="sd">        local_p_df.columns = pair_names</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="n">L_mtx</span> <span class="o">=</span> <span class="n">eSP2</span><span class="o">.</span><span class="n">associations_</span>
    <span class="n">pair_names</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;~&#39;</span> <span class="o">+</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">peaks</span><span class="p">)</span>
    <span class="n">pair_names</span> <span class="o">=</span> <span class="n">pair_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">L_mtx</span><span class="p">,</span> <span class="n">pair_names</span>




<div class="viewcode-block" id="Local_L"><a class="viewcode-back" href="../../api.html#treasmo.core.Local_L">[docs]</a><span class="k">def</span> <span class="nf">Local_L</span><span class="p">(</span><span class="n">mudata</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">peaks</span><span class="p">,</span>
            <span class="n">mods</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">,</span><span class="s1">&#39;atac&#39;</span><span class="p">],</span>
            <span class="n">rm_dropout</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_RAM</span><span class="o">=</span><span class="mi">16</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    Function to calculate the single-cell gene-peak correlation strength index for all the pairs in multiome data</span>

<span class="sd">    Arguments</span>
<span class="sd">    --------------</span>

<span class="sd">        mudata: MuData</span>
<span class="sd">            single-cell multi-omics data saved as MuData object</span>

<span class="sd">        mods: List[str, str]</span>
<span class="sd">            scRNA-seq and scATAC-seq modality name in MuData object</span>

<span class="sd">        genes, peaks: List, numpy.array</span>
<span class="sd">            one-to-one lists containing gene and peak names</span>

<span class="sd">        rm_dropout: bool</span>
<span class="sd">            make correlation strength index to be 0 if feature value is 0 (dropout)</span>

<span class="sd">        seed: int</span>
<span class="sd">            random seed to make the results reproducible</span>

<span class="sd">        max_RAM: int</span>
<span class="sd">            maximum limitation of memory usage (Gb)</span>


<span class="sd">    Returns</span>
<span class="sd">    --------------</span>
<span class="sd">        MuData</span>
<span class="sd">            added with Local L matrix and gene-peak pair names in ``mudata.uns``</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Inferring KNN graph-based local correlation...&quot;</span><span class="p">)</span>
    
    <span class="n">L_mtx</span><span class="p">,</span> <span class="n">L_mtx_names</span> <span class="o">=</span> <span class="n">_calculate_LL</span><span class="p">(</span><span class="n">mudata</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">peaks</span><span class="p">,</span>
                                       <span class="n">mods</span><span class="o">=</span><span class="n">mods</span><span class="p">,</span>
                                       <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_RAM</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">rm_dropout</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Setting local correlation to 0 for cells with no expression on either feature of a certain pair...&quot;</span><span class="p">)</span>
        <span class="n">GP_names</span> <span class="o">=</span> <span class="n">L_mtx_names</span>
        <span class="n">Dropout_mtx</span> <span class="o">=</span> <span class="n">_dropout_filter</span><span class="p">(</span><span class="n">mudata</span><span class="p">,</span> <span class="n">GP_names</span><span class="p">,</span> <span class="n">mods</span><span class="p">)</span>
        <span class="n">L_mtx</span> <span class="o">=</span> <span class="n">L_mtx</span> <span class="o">*</span> <span class="n">Dropout_mtx</span>

    <span class="n">mudata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;Local_L&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">L_mtx</span>
    <span class="n">mudata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;Local_L_names&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">L_mtx_names</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Following changes made to the AnnData object:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">KNN graph-based local correlation results saved in uns[&#39;Local_L&#39;]&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">Gene-peak pair names saved in uns[&#39;Local_L_names&#39;]&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Finished Inferring local correlation. </span><span class="si">%.3f</span><span class="s2">s past&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">mudata</span></div>




<span class="k">def</span> <span class="nf">_dropout_filter</span><span class="p">(</span><span class="n">mudata</span><span class="p">,</span> <span class="n">GP_names</span><span class="p">,</span> <span class="n">mods</span><span class="p">):</span>

    <span class="c1"># Gene dropout</span>
    <span class="n">index_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;index&#39;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">var_names</span><span class="p">))})</span>
    <span class="n">index_df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">var_names</span>
    <span class="n">GP_G</span> <span class="o">=</span> <span class="p">[</span><span class="n">gp</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;~&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">gp</span> <span class="ow">in</span> <span class="n">GP_names</span><span class="p">]</span>
    <span class="n">GP_genes_index</span> <span class="o">=</span> <span class="n">index_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">GP_G</span><span class="p">,:][</span><span class="s1">&#39;index&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

    <span class="n">E_mtx</span> <span class="o">=</span> <span class="n">_check_array</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
    <span class="n">E_mtx_dropout_value</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">E_mtx</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="n">Dropout_mtx</span> <span class="o">=</span> <span class="p">(</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isclose</span><span class="p">(</span><span class="n">E_mtx_dropout_value</span><span class="p">,</span><span class="n">E_mtx</span><span class="p">,</span> <span class="mf">1e-3</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">Dropout_mtx_G</span> <span class="o">=</span> <span class="n">Dropout_mtx</span><span class="p">[:,</span><span class="n">GP_genes_index</span><span class="p">]</span>

    <span class="c1"># Peak dropout</span>
    <span class="n">index_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;index&#39;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">var_names</span><span class="p">))})</span>
    <span class="n">index_df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">var_names</span>
    <span class="n">GP_P</span> <span class="o">=</span> <span class="p">[</span><span class="n">gp</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;~&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">gp</span> <span class="ow">in</span> <span class="n">GP_names</span><span class="p">]</span>
    <span class="n">GP_peaks_index</span> <span class="o">=</span> <span class="n">index_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">GP_P</span><span class="p">,:][</span><span class="s1">&#39;index&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

    <span class="n">E_mtx</span> <span class="o">=</span> <span class="n">_check_array</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
    <span class="n">E_mtx_dropout_value</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">E_mtx</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="n">Dropout_mtx</span> <span class="o">=</span> <span class="p">(</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isclose</span><span class="p">(</span><span class="n">E_mtx_dropout_value</span><span class="p">,</span><span class="n">E_mtx</span><span class="p">,</span> <span class="mf">1e-3</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">Dropout_mtx_P</span> <span class="o">=</span> <span class="n">Dropout_mtx</span><span class="p">[:,</span><span class="n">GP_peaks_index</span><span class="p">]</span>
    
    <span class="n">Dropout_mtx</span> <span class="o">=</span> <span class="n">Dropout_mtx_G</span> <span class="o">*</span> <span class="n">Dropout_mtx_P</span>
    
    <span class="k">return</span> <span class="n">Dropout_mtx</span>





<span class="c1">#=========================================================================================</span>
<span class="c1"># Pearson correlation function for benchmark purpose</span>
<span class="c1">#=========================================================================================</span>

<div class="viewcode-block" id="Pearsonr"><a class="viewcode-back" href="../../api.html#treasmo.core.Pearsonr">[docs]</a><span class="k">def</span> <span class="nf">Pearsonr</span><span class="p">(</span><span class="n">mudata</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">peaks</span><span class="p">,</span>
             <span class="n">mods</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">,</span><span class="s1">&#39;atac&#39;</span><span class="p">],</span>
             <span class="n">p_value</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="w">    </span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to calculate the Pearson correlation between genes and peaks</span>

<span class="sd">    Arguments</span>
<span class="sd">    --------------</span>

<span class="sd">        mudata: MuData</span>
<span class="sd">            single-cell multi-omics data saved as MuData object</span>

<span class="sd">        mods: List[str, str]</span>
<span class="sd">            scRNA-seq and scATAC-seq modality name in MuData object</span>

<span class="sd">        genes, peaks: List, numpy.array</span>
<span class="sd">            one-to-one lists containing gene and peak names</span>

<span class="sd">        p_value: bool</span>
<span class="sd">            perform significance testing or not</span>

<span class="sd">        seed: int</span>
<span class="sd">            random seed to make the results reproducible</span>


<span class="sd">    Returns</span>
<span class="sd">    --------------</span>
<span class="sd">        DataFrame</span>
<span class="sd">            containing Pearson correlation r, p-value and FDR for all gene-peak pairs</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>

    <span class="n">gene_X</span> <span class="o">=</span> <span class="n">_prepare_mtx</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">genes</span><span class="p">)</span>
    <span class="n">peak_X</span> <span class="o">=</span> <span class="n">_prepare_mtx</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mods</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span> <span class="n">peaks</span><span class="p">)</span>

    <span class="n">gene_X</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_X</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">gene_X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">columns</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">gene_X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="n">peak_X</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">peak_X</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">peak_X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">columns</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">gene_X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

    <span class="n">global_P_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">peak_X</span><span class="o">.</span><span class="n">corrwith</span><span class="p">(</span><span class="n">gene_X</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
    <span class="n">global_P_df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;~&#39;</span> <span class="o">+</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">peaks</span><span class="p">)</span>
    <span class="n">global_P_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;r&#39;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">p_value</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Calculating p-values...&quot;</span><span class="p">)</span>
        <span class="n">p_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">global_P_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="n">p_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">stats</span><span class="o">.</span><span class="n">pearsonr</span><span class="p">(</span><span class="n">gene_X</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">peak_X</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">global_P_df</span><span class="p">[</span><span class="s1">&#39;r.p_value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">p_list</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Calculating FDR...&quot;</span><span class="p">)</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">global_P_df</span><span class="p">[</span><span class="s1">&#39;r.FDR&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fdrcorrection</span><span class="p">(</span><span class="n">global_P_df</span><span class="p">[</span><span class="s1">&#39;r.p_value&#39;</span><span class="p">],</span>
                                                <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;indep&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">global_P_df</span><span class="p">[</span><span class="s1">&#39;r.p_value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">NaN</span>
        <span class="n">global_P_df</span><span class="p">[</span><span class="s1">&#39;r.FDR&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">NaN</span>

    <span class="k">return</span> <span class="n">global_P_df</span></div>







</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Chaozhong Liu
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2023, Zhandong Liu&#39;s Lab at Baylor College of Medicine.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=365ca57ee442770a23c6"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=365ca57ee442770a23c6"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>