

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>treasmo.tl &#8212; TREASMO v1.1.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=365ca57ee442770a23c6" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=365ca57ee442770a23c6" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=365ca57ee442770a23c6" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=365ca57ee442770a23c6"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/treasmo/tl';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">TREASMO v1.1.0 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">



<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <h1>Source code for treasmo.tl</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>

<span class="kn">import</span> <span class="nn">scipy</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">time</span>

<span class="kn">from</span> <span class="nn">sklearn.neighbors</span> <span class="kn">import</span> <span class="n">KNeighborsRegressor</span>


<span class="n">__author__</span> <span class="o">=</span> <span class="s2">&quot;Chaozhong Liu&quot;</span>
<span class="n">__email__</span> <span class="o">=</span> <span class="s2">&quot;czliubioinfo@gmail.com&quot;</span>


<span class="c1"># ============= Toolkit functions =============</span>
<span class="c1"># For Future work</span>
<span class="c1">#   Pseudo-bulk Generation (and the functions mapping results back to single-cell data)</span>
<span class="c1">#   nearest neighbor map construction</span>
<span class="c1">#   imputation option</span>


<span class="c1"># Implemented function</span>
<span class="c1">#   Feature sparsity summary</span>
<span class="c1">#   Gene-peak annotation</span>
<span class="c1">#   TF-peak annotation</span>
<span class="c1"># =============================================</span>



<span class="c1"># Calculate feature sparsity =================================================</span>

<span class="k">def</span> <span class="nf">_fill_miss_index</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Internal helper function for feature sparsity calculation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">miss_ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">setdiff1d</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">n</span><span class="p">),</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;ind&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>
    <span class="n">miss_pd</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;ind&#39;</span><span class="p">:</span><span class="n">miss_ind</span><span class="p">,</span><span class="s1">&#39;size&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">})</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="p">,</span><span class="n">miss_pd</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;ind&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span>


<div class="viewcode-block" id="feature_sparsity"><a class="viewcode-back" href="../../api.html#treasmo.tl.feature_sparsity">[docs]</a><span class="k">def</span> <span class="nf">feature_sparsity</span><span class="p">(</span><span class="n">mudata</span><span class="p">,</span> <span class="n">group_by</span><span class="o">=</span><span class="p">{}):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to add feature sparisity information in MuData.var</span>
<span class="sd">    The value is defined as ``#non-zero value / #cells`` in each feature.</span>

<span class="sd">    Arguments</span>
<span class="sd">    --------------</span>
<span class="sd">        mudata: MuData</span>
<span class="sd">            single-cell multi-omics data saved as MuData object</span>

<span class="sd">        group_by: dict</span>
<span class="sd">            If provided, calculate per group feature sparsity for each modality</span>

<span class="sd">            Example: {&#39;rna&#39;:&#39;cell_type&#39;, &#39;atac&#39;:&#39;cell_type&#39;}</span>

<span class="sd">    Return</span>
<span class="sd">    --------------</span>
<span class="sd">        MuData</span>
<span class="sd">            var[&#39;Frac.all&#39;]</span>
<span class="sd">            var[&#39;Frac.GroupName&#39;]</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">mod</span> <span class="ow">in</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>

        <span class="n">spsmtx</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
        <span class="n">non_zero</span> <span class="o">=</span> <span class="n">spsmtx</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()</span>
        <span class="n">non_zero_row</span> <span class="o">=</span> <span class="n">non_zero</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">non_zero_ind</span> <span class="o">=</span> <span class="n">non_zero</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="c1">## All cell sparsity</span>
        <span class="n">non_zero_pd</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">non_zero_ind</span><span class="p">)</span>
        <span class="n">non_zero_pd</span> <span class="o">=</span> <span class="n">non_zero_pd</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
        <span class="n">non_zero_pd</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">non_zero_pd</span><span class="p">)</span>
        <span class="n">non_zero_pd</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ind&#39;</span><span class="p">,</span><span class="s1">&#39;size&#39;</span><span class="p">]</span>
        <span class="n">non_zero_pd</span> <span class="o">=</span> <span class="n">_fill_miss_index</span><span class="p">(</span><span class="n">non_zero_pd</span><span class="p">,</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">n_vars</span><span class="p">)</span>
        <span class="n">non_zero_pd</span><span class="p">[</span><span class="s1">&#39;sparsity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">non_zero_pd</span><span class="p">[</span><span class="s1">&#39;size&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">n_obs</span>

        <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;Frac.all&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">non_zero_pd</span><span class="p">[</span><span class="s1">&#39;sparsity&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

        <span class="c1">## Per group sparsity</span>
        <span class="k">if</span> <span class="n">mod</span> <span class="ow">in</span> <span class="n">group_by</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>

            <span class="n">non_zero_pd</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;ind&#39;</span><span class="p">:</span><span class="n">non_zero_ind</span><span class="p">,</span>
                                        <span class="s1">&#39;group&#39;</span><span class="p">:</span><span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">group_by</span><span class="p">[</span><span class="n">mod</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()[</span><span class="n">non_zero_row</span><span class="p">]})</span>
            <span class="n">non_zero_pd</span> <span class="o">=</span> <span class="n">non_zero_pd</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ind&#39;</span><span class="p">,</span><span class="s1">&#39;group&#39;</span><span class="p">],</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;group&#39;</span><span class="p">)</span>
            <span class="n">non_zero_pd</span><span class="p">[</span><span class="s1">&#39;n_cells&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">group_by</span><span class="p">[</span><span class="n">mod</span><span class="p">]]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()[</span><span class="n">non_zero_pd</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

            <span class="k">for</span> <span class="n">ident</span> <span class="ow">in</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">group_by</span><span class="p">[</span><span class="n">mod</span><span class="p">]]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">non_zero_pd</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">non_zero_pd</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">ident</span><span class="p">,:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                <span class="n">n_cells</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;n_cells&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,[</span><span class="s1">&#39;ind&#39;</span><span class="p">,</span><span class="s1">&#39;size&#39;</span><span class="p">]]</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">_fill_miss_index</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">n_vars</span><span class="p">)</span>
                <span class="n">df</span><span class="p">[</span><span class="s1">&#39;sparsity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;size&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">n_cells</span>
                <span class="n">mudata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;Frac.</span><span class="si">%s</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">ident</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;sparsity&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>


    <span class="k">return</span> <span class="n">mudata</span></div>




<span class="c1"># Annotate genes with nearby peaks =================================================</span>

<div class="viewcode-block" id="get_gloc_from_atac_data"><a class="viewcode-back" href="../../api.html#treasmo.tl.get_gloc_from_atac_data">[docs]</a><span class="k">def</span> <span class="nf">get_gloc_from_atac_data</span><span class="p">(</span><span class="n">peaks</span><span class="p">,</span> <span class="n">split_symbol</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function to get the genomic locations (including the middle point)of peaks</span>
<span class="sd">        Author: Linhua Wang Linhua.Wang@bcm.edu</span>
<span class="sd">        https://github.com/LiuzLab/Neurips2021/blob/master/task1_utils1.py</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">glocs</span> <span class="o">=</span> <span class="n">peaks</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">glocs</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">glocs</span> <span class="k">if</span> <span class="s1">&#39;chr&#39;</span> <span class="ow">in</span> <span class="n">c</span><span class="p">]</span>
    <span class="n">chrms</span><span class="p">,</span> <span class="n">ranges</span><span class="p">,</span> <span class="n">sts</span><span class="p">,</span> <span class="n">ends</span><span class="p">,</span> <span class="n">midpoints</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">gl</span> <span class="ow">in</span> <span class="n">glocs</span><span class="p">:</span>
        <span class="n">chrms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gl</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">st</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gl</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">gl</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">sts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">st</span><span class="p">)</span>
        <span class="n">ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">end</span><span class="p">)</span>
        <span class="c1">#midpoints.append(int((st + end)/2))</span>
        <span class="c1">#ranges.append(&quot;_&quot;.join(gl.split(split_symbol[0])[1].split(split_symbol[1])))</span>
    <span class="n">gloc_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrm&#39;</span><span class="p">:</span> <span class="n">chrms</span><span class="p">,</span> <span class="c1">#&#39;grange&#39;: ranges,</span>
                        <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="n">sts</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="n">ends</span><span class="p">},</span> <span class="n">index</span><span class="o">=</span><span class="n">glocs</span><span class="p">)</span>
                        <span class="c1">#&#39;midpoint&#39;: midpoints}, index=glocs)</span>
    <span class="k">return</span> <span class="n">gloc_df</span></div>




<div class="viewcode-block" id="nearby_peaks"><a class="viewcode-back" href="../../api.html#treasmo.tl.nearby_peaks">[docs]</a><span class="k">def</span> <span class="nf">nearby_peaks</span><span class="p">(</span><span class="n">g_array</span><span class="p">,</span> <span class="n">min_op</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function to get regions overlapped with the gene body.</span>

<span class="sd">    Arguments</span>
<span class="sd">    --------------</span>
<span class="sd">        min_op: int</span>
<span class="sd">            The peak and gene have to be at least ``min_op`` overlapped to be considered overlapped.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">g_chr</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">g_array</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
    <span class="n">g_array_n</span> <span class="o">=</span> <span class="n">g_array</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">op</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">minimum</span><span class="p">(</span><span class="n">p_array</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">squeeze</span><span class="p">(),</span><span class="n">g_array_n</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">p_array</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">squeeze</span><span class="p">(),</span><span class="n">g_array_n</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">filter_bool</span> <span class="o">=</span> <span class="p">(</span><span class="n">op</span><span class="o">&gt;=</span><span class="n">min_op</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">chr_list</span><span class="o">==</span><span class="n">g_chr</span><span class="p">)</span>
    <span class="n">op_index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">op</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])[</span><span class="n">filter_bool</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">plist</span><span class="p">[</span><span class="n">op_index</span><span class="p">])),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span></div>
    

<div class="viewcode-block" id="peaks_within_distance"><a class="viewcode-back" href="../../api.html#treasmo.tl.peaks_within_distance">[docs]</a><span class="k">def</span> <span class="nf">peaks_within_distance</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">peaks</span><span class="p">,</span> <span class="n">upstream</span><span class="p">,</span> <span class="n">downstream</span><span class="p">,</span> <span class="n">ref_gtf_fn</span><span class="p">,</span>
                          <span class="n">no_intersect</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">id_col</span><span class="o">=</span><span class="s1">&#39;GeneSymbol&#39;</span><span class="p">,</span> <span class="n">split_symbol</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;-&#39;</span><span class="p">]):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to annotate genes with nearby peaks</span>

<span class="sd">    Arguments</span>
<span class="sd">    --------------</span>
<span class="sd">        genes: List, numpy.array</span>
<span class="sd">            gene list to be annotated</span>

<span class="sd">        peaks: List, numpy.array</span>
<span class="sd">            candidate peaks list</span>

<span class="sd">        upstream: int</span>
<span class="sd">            include peaks N bp upstream of the gene TSS</span>

<span class="sd">        downstream: int</span>
<span class="sd">            include peaks N bp downstream of the TES</span>

<span class="sd">        ref_gtf_fn: str</span>
<span class="sd">            GTF format file containing gene location information</span>
<span class="sd">                see example at https://github.com/ChaozhongLiu/scGREAT/tree/main/replication/data</span>
<span class="sd">                - Homo_sapiens.GRCh38.104.GeneLoc.Tab.txt</span>
<span class="sd">                - Mus_musculus.GRCm38.100.GeneLoc.Tab.txt</span>

<span class="sd">        no_intersect: bool</span>
<span class="sd">            if the candidate peak lies in another gene&#39;s body, remove the peak or not</span>

<span class="sd">        id_col: str</span>
<span class="sd">            column name in ``ref_gtf_fn`` that indicates gene ID</span>

<span class="sd">        split_symbol: List[str, str]</span>
<span class="sd">            how peak location ID is formatted</span>
<span class="sd">                &#39;chr1-12345-23456&#39; - split_symbol=[&#39;-&#39;,&#39;-&#39;]</span>
<span class="sd">                &#39;chr1:12345-23456&#39; - split_symbol=[&#39;:&#39;,&#39;-&#39;]</span>

<span class="sd">    Return</span>
<span class="sd">    --------------</span>
<span class="sd">        DataFrame</span>
<span class="sd">            contains gene annotation information</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">gloc_df</span> <span class="o">=</span> <span class="n">get_gloc_from_atac_data</span><span class="p">(</span><span class="n">peaks</span><span class="p">,</span> <span class="n">split_symbol</span><span class="o">=</span><span class="n">split_symbol</span><span class="p">)</span>
    
    <span class="n">ref_gtf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">ref_gtf_fn</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="c1">#if id_col==&#39;GeneSymbol&#39;:</span>
    <span class="c1">#    ref_gtf = ref_gtf.loc[ref_gtf[&#39;GeneSymbol&#39;].isin(genes)]</span>
    <span class="c1">#elif id_col==&#39;gene_id&#39;:</span>
    <span class="n">ref_gtf</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="n">id_col</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">genes</span><span class="p">)]</span>
    <span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;GeneSymbol&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="p">[</span><span class="n">id_col</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1">#upstream, downstream = 100000, 100000</span>
    <span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;start_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;end_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;+&#39;</span><span class="p">,</span><span class="s1">&#39;start_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;+&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">upstream</span>
    <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;start_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">downstream</span>

    <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;+&#39;</span><span class="p">,</span><span class="s1">&#39;end_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;+&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">downstream</span>
    <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;end_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">upstream</span>
    
    <span class="k">global</span> <span class="n">p_array</span><span class="p">,</span> <span class="n">plist</span><span class="p">,</span> <span class="n">chr_list</span>
    <span class="n">p_array</span> <span class="o">=</span> <span class="n">gloc_df</span><span class="p">[[</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">plist</span> <span class="o">=</span> <span class="n">gloc_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">chr_list</span> <span class="o">=</span> <span class="n">gloc_df</span><span class="p">[</span><span class="s1">&#39;chrm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">object</span><span class="p">)</span>

    <span class="n">genes_loc</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="p">[[</span><span class="s1">&#39;start_exp&#39;</span><span class="p">,</span><span class="s1">&#39;end_exp&#39;</span><span class="p">,</span><span class="s1">&#39;chr&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">selected_peaks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">apply_along_axis</span><span class="p">(</span><span class="n">nearby_peaks</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">arr</span><span class="o">=</span><span class="n">genes_loc</span><span class="p">)</span>
    
    <span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;nearby_peaks&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">selected_peaks</span>
    <span class="n">ref_gtf</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="p">(</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;nearby_peaks&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;&#39;</span><span class="p">)]</span>
    <span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;nearby_peaks&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;nearby_peaks&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
    
    <span class="n">peaks_nearby_new</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">6</span><span class="p">)),</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GeneSymbol&#39;</span><span class="p">,</span><span class="s1">&#39;Strand&#39;</span><span class="p">,</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">,</span><span class="s1">&#39;nearby_peaks&#39;</span><span class="p">])</span> <span class="c1">#ref_gtf.columns[[1,4,5,6,7,10]])</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ref_gtf</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">n_lines</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="s1">&#39;nearby_peaks&#39;</span><span class="p">])</span>
        <span class="n">peaks_nearby_new_tmp</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">n_lines</span><span class="p">),:]</span>
        <span class="n">peaks_nearby_new_tmp</span> <span class="o">=</span> <span class="n">peaks_nearby_new_tmp</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,[</span><span class="s1">&#39;GeneSymbol&#39;</span><span class="p">,</span><span class="s1">&#39;Strand&#39;</span><span class="p">,</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">,</span><span class="s1">&#39;nearby_peaks&#39;</span><span class="p">]]</span>
        <span class="n">peaks_nearby_new_tmp</span><span class="p">[</span><span class="s1">&#39;nearby_peaks&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="s1">&#39;nearby_peaks&#39;</span><span class="p">]</span>

        <span class="n">peaks_nearby_new</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">peaks_nearby_new</span><span class="p">,</span> <span class="n">peaks_nearby_new_tmp</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    
    <span class="c1">#return peaks_nearby_new</span>

    <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;midp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;nearby_peaks&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> \
                               <span class="nb">int</span><span class="p">((</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span> <span class="p">)</span>

    <span class="n">plus_strand</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;+&#39;</span>

    <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;tss_dist&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">NaN</span>
    <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;tss_dist&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;midp&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">]</span>
    <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;tss_dist&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;midp&#39;</span><span class="p">]</span>

    <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;tts_dist&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">NaN</span>
    <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;tts_dist&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;midp&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">]</span>
    <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;tts_dist&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">plus_strand</span><span class="p">,</span><span class="s1">&#39;midp&#39;</span><span class="p">]</span>

    <span class="n">promoters</span> <span class="o">=</span> <span class="p">((</span><span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;tss_dist&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="o">-</span><span class="mi">2000</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;tss_dist&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">genebodys</span> <span class="o">=</span> <span class="p">((</span><span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;tss_dist&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;tts_dist&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

    <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;pRegion&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">promoters</span>
    <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;gBody&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genebodys</span>
    
    <span class="n">peaks_nearby_final</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;tss_dist&#39;</span><span class="p">]</span><span class="o">&gt;=-</span><span class="n">upstream</span><span class="p">)</span><span class="o">&amp;</span>
                                              <span class="p">(</span><span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;tts_dist&#39;</span><span class="p">]</span><span class="o">&lt;=</span><span class="n">downstream</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    
    <span class="n">df</span> <span class="o">=</span> <span class="n">peaks_nearby_final</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span><span class="s1">&#39;Strand&#39;</span><span class="p">,</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">,</span><span class="s1">&#39;Peak&#39;</span><span class="p">,</span><span class="s1">&#39;midp&#39;</span><span class="p">,</span><span class="s1">&#39;tss_dist&#39;</span><span class="p">,</span><span class="s1">&#39;tts_dist&#39;</span><span class="p">,</span><span class="s1">&#39;pRegion&#39;</span><span class="p">,</span><span class="s1">&#39;gBody&#39;</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,[</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">]]</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(),:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    
    <span class="k">if</span> <span class="n">no_intersect</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Remove nearby peaks if it lies on the gene body or promoter regions of other genes.&quot;</span><span class="p">)</span>
        <span class="n">duplicated_peaks</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Peak&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">dup_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">duplicated_peaks</span><span class="p">,:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">dup_df</span> <span class="o">=</span> <span class="n">dup_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;Peak&#39;</span><span class="p">)[[</span><span class="s1">&#39;pRegion&#39;</span><span class="p">,</span><span class="s1">&#39;gBody&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
        <span class="n">body_peaks</span> <span class="o">=</span> <span class="n">dup_df</span><span class="o">.</span><span class="n">index</span><span class="p">[(</span><span class="n">dup_df</span><span class="p">[</span><span class="s1">&#39;pRegion&#39;</span><span class="p">]</span><span class="o">==</span><span class="mi">1</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">dup_df</span><span class="p">[</span><span class="s1">&#39;gBody&#39;</span><span class="p">]</span><span class="o">==</span><span class="mi">1</span><span class="p">)]</span>

        <span class="n">peaks_rm</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Peak&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">body_peaks</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;pRegion&#39;</span><span class="p">]</span><span class="o">==</span><span class="mi">0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;gBody&#39;</span><span class="p">]</span><span class="o">==</span><span class="mi">0</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">peaks_rm</span><span class="p">,:]</span>
    
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,[</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">10</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    
    <span class="k">return</span> <span class="n">df</span></div>




<span class="c1"># Annotate TF with motif regions =================================================</span>


<div class="viewcode-block" id="TFBS_match"><a class="viewcode-back" href="../../api.html#treasmo.tl.TFBS_match">[docs]</a><span class="k">def</span> <span class="nf">TFBS_match</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">peaks</span><span class="p">,</span> <span class="n">ref_fn</span><span class="p">,</span> <span class="n">min_overlap</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">split_symbol</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;-&#39;</span><span class="p">]):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to annotate TF with binding site regions</span>

<span class="sd">    Arguments</span>
<span class="sd">    --------------</span>
<span class="sd">        genes: List, numpy.array</span>
<span class="sd">            gene list to be annotated</span>

<span class="sd">        peaks: List, numpy.array</span>
<span class="sd">            candidate peaks list</span>

<span class="sd">        ref_fn: str</span>
<span class="sd">            BED format file containing gene binding site location information</span>
<span class="sd">            see example below from JASPAR TFBS genome track (https://jaspar.genereg.net/genome-tracks/)</span>
<span class="sd">                chr1    280     298     AGL3    821     -</span>
<span class="sd">                chr1    309     327     AGL3    823     +</span>
<span class="sd">                chr1    309     327     AGL3    882     -</span>
<span class="sd">                chr1    1577    1595    AGL3    823     +</span>
<span class="sd">        </span>
<span class="sd">        min_overlap: int</span>
<span class="sd">            the minimum number of overlaped base pairs between candidate peak and TFBS</span>

<span class="sd">        split_symbol: List[str, str]</span>
<span class="sd">            how peak location ID is formatted</span>
<span class="sd">                &#39;chr1-12345-23456&#39; - split_symbol=[&#39;-&#39;,&#39;-&#39;]</span>
<span class="sd">                &#39;chr1:12345-23456&#39; - split_symbol=[&#39;:&#39;,&#39;-&#39;]</span>

<span class="sd">    Return</span>
<span class="sd">    --------------</span>
<span class="sd">        DataFrame</span>
<span class="sd">            contains TF annotation information</span>

<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">gloc_df</span> <span class="o">=</span> <span class="n">get_gloc_from_atac_data</span><span class="p">(</span><span class="n">peaks</span><span class="p">,</span> <span class="n">split_symbol</span><span class="o">=</span><span class="n">split_symbol</span><span class="p">)</span>
    
    <span class="n">ref_gtf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">ref_fn</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">ref_gtf</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;score&#39;</span><span class="p">,</span> <span class="s1">&#39;Strand&#39;</span><span class="p">]</span>
    <span class="n">ref_gtf</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">genes</span><span class="p">)]</span>

    <span class="n">upstream</span><span class="p">,</span> <span class="n">downstream</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
    <span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;start_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;end_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;+&#39;</span><span class="p">,</span><span class="s1">&#39;start_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;+&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">upstream</span>
    <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;start_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">downstream</span>

    <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;+&#39;</span><span class="p">,</span><span class="s1">&#39;end_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;+&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">downstream</span>
    <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;end_exp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">upstream</span>

    <span class="k">global</span> <span class="n">p_array</span><span class="p">,</span> <span class="n">plist</span><span class="p">,</span> <span class="n">chr_list</span>
    <span class="n">p_array</span> <span class="o">=</span> <span class="n">gloc_df</span><span class="p">[[</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">plist</span> <span class="o">=</span> <span class="n">gloc_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">chr_list</span> <span class="o">=</span> <span class="n">gloc_df</span><span class="p">[</span><span class="s1">&#39;chrm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">object</span><span class="p">)</span>

    <span class="n">genes_loc</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="p">[[</span><span class="s1">&#39;start_exp&#39;</span><span class="p">,</span><span class="s1">&#39;end_exp&#39;</span><span class="p">,</span><span class="s1">&#39;chr&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">selected_peaks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">apply_along_axis</span><span class="p">(</span><span class="n">nearby_peaks</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">arr</span><span class="o">=</span><span class="n">genes_loc</span><span class="p">)</span>
    
    <span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;TFBS&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">selected_peaks</span>
    <span class="n">ref_gtf</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="p">(</span><span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;TFBS&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;&#39;</span><span class="p">)]</span>
    <span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;TFBS&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="p">[</span><span class="s1">&#39;TFBS&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>

    <span class="n">peaks_nearby_new</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">6</span><span class="p">)),</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span><span class="s1">&#39;Strand&#39;</span><span class="p">,</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">,</span><span class="s1">&#39;TFBS&#39;</span><span class="p">])</span> <span class="c1">#ref_gtf.columns[[1,4,5,6,7,10]])</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">ref_gtf</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">n_lines</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="s1">&#39;TFBS&#39;</span><span class="p">])</span>
        <span class="n">peaks_nearby_new_tmp</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">n_lines</span><span class="p">),:]</span>
        <span class="n">peaks_nearby_new_tmp</span> <span class="o">=</span> <span class="n">peaks_nearby_new_tmp</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span><span class="s1">&#39;Strand&#39;</span><span class="p">,</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">,</span><span class="s1">&#39;TFBS&#39;</span><span class="p">]]</span>
        <span class="n">peaks_nearby_new_tmp</span><span class="p">[</span><span class="s1">&#39;TFBS&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_gtf</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="s1">&#39;TFBS&#39;</span><span class="p">]</span>

        <span class="n">peaks_nearby_new</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">peaks_nearby_new</span><span class="p">,</span> <span class="n">peaks_nearby_new_tmp</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    
    <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;pstart&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;TFBS&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;pstart&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;pstart&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;pend&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;TFBS&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;pend&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;pend&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>

    <span class="n">peaks_nearby_new</span><span class="p">[</span><span class="s1">&#39;overlap&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;pend&#39;</span><span class="p">],</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">])</span> <span class="o">-</span> <span class="nb">max</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;pstart&#39;</span><span class="p">],</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">df</span> <span class="o">=</span> <span class="n">peaks_nearby_new</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s1">&#39;pstart&#39;</span><span class="p">,</span> <span class="s1">&#39;pend&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span><span class="s1">&#39;Strand&#39;</span><span class="p">,</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">,</span><span class="s1">&#39;Peak&#39;</span><span class="p">,</span><span class="s1">&#39;overlap&#39;</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;overlap&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,[</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">]]</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(),:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    
    <span class="c1">#df = df.iloc[:,[0,5,7,9,8,10]].copy()</span>
    <span class="n">df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    
    <span class="k">return</span> <span class="n">df</span></div>



<span class="c1">#=========================================================================================</span>
<span class="c1"># Homer related tools</span>
<span class="c1">#=========================================================================================</span>
<div class="viewcode-block" id="PFM2Motif"><a class="viewcode-back" href="../../api.html#treasmo.tl.PFM2Motif">[docs]</a><span class="k">def</span> <span class="nf">PFM2Motif</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">out_file</span><span class="p">,</span> <span class="n">detection_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to convert PFM (Position Frequency Matrix) into Homer Motif file</span>

<span class="sd">    Arguments</span>
<span class="sd">    --------------</span>
<span class="sd">        file_name: str</span>
<span class="sd">            Path to PFM file, see example at https://jaspar.elixir.no/docs/#jaspar-matrix-formats</span>
<span class="sd">        </span>
<span class="sd">        out_file: str</span>
<span class="sd">            Path to output file</span>
<span class="sd">        </span>
<span class="sd">        detection_threshold: int, float</span>
<span class="sd">            See Homer explanation at http://homer.ucsd.edu/homer/motif/creatingCustomMotifs.html</span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>

    <span class="n">df_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">5</span><span class="p">):</span>
        <span class="n">line</span> <span class="o">=</span> <span class="n">file</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
        <span class="n">line</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">a</span><span class="p">:</span> <span class="n">a</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">line</span><span class="p">))</span>
        <span class="n">line</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s1">&#39;[&#39;</span><span class="p">)</span>
        <span class="n">line</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s1">&#39;]&#39;</span><span class="p">)</span>
        <span class="c1">#line = line[1:]</span>
        <span class="n">df_dict</span><span class="p">[</span><span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">:]]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">df_dict</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span><span class="s1">&#39;C&#39;</span><span class="p">,</span><span class="s1">&#39;G&#39;</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">]]</span>
    <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span> <span class="o">/</span> <span class="n">df</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]</span>

    <span class="n">motif_seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)])</span>
    <span class="n">motif_header</span> <span class="o">=</span> <span class="n">file</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">motif_header</span> <span class="o">=</span> <span class="n">motif_header</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)</span>
    <span class="n">motif_header</span> <span class="o">=</span> <span class="n">motif_header</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">motif_header</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">motif_header</span><span class="p">)</span>
    <span class="n">motif_header</span> <span class="o">=</span> <span class="s1">&#39;&gt;&#39;</span> <span class="o">+</span> <span class="n">motif_seq</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">motif_header</span> <span class="o">+</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="se">\t</span><span class="si">{</span><span class="n">detection_threshold</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="n">motif_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_file</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
    <span class="n">motif_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">motif_header</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">freq_list</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">,:]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">freq_list</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">j</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">freq_list</span><span class="p">]</span>
        <span class="n">motif_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">motif_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">freq_list</span><span class="p">))</span>
        
    <span class="n">motif_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>




<div class="viewcode-block" id="peak2HomerInput"><a class="viewcode-back" href="../../api.html#treasmo.tl.peak2HomerInput">[docs]</a><span class="k">def</span> <span class="nf">peak2HomerInput</span><span class="p">(</span><span class="n">peaks</span><span class="p">,</span> <span class="n">out_file</span><span class="p">,</span> <span class="n">filetype</span><span class="o">=</span><span class="s1">&#39;peaks&#39;</span><span class="p">,</span> <span class="n">split_symbol</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;-&#39;</span><span class="p">]):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function to convert peak list to Homer accepted input format</span>

<span class="sd">    Arguments</span>
<span class="sd">    --------------</span>
<span class="sd">        peaks: List, numpy.array</span>
<span class="sd">            peaks list to convert</span>
<span class="sd">        </span>
<span class="sd">        out_file: str</span>
<span class="sd">            Path to output file</span>
<span class="sd">        </span>
<span class="sd">        filetype: str</span>
<span class="sd">            peaks or bed. File type to generate</span>
<span class="sd">        </span>
<span class="sd">        split_symbol: List[str, str]</span>
<span class="sd">            how peak location ID is formatted</span>
<span class="sd">                &#39;chr1-12345-23456&#39; - split_symbol=[&#39;-&#39;,&#39;-&#39;]</span>
<span class="sd">                &#39;chr1:12345-23456&#39; - split_symbol=[&#39;:&#39;,&#39;-&#39;]</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">filetype</span> <span class="o">==</span> <span class="s1">&#39;peaks&#39;</span><span class="p">:</span>
        <span class="n">homer_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;index&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">peaks</span><span class="p">)),</span>
                                 <span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">],</span>
                                 <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">],</span>
                                 <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">])[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">],</span>
                                 <span class="s1">&#39;strand&#39;</span><span class="p">:</span><span class="s1">&#39;.&#39;</span><span class="p">})</span>

    <span class="k">elif</span> <span class="n">filetype</span> <span class="o">==</span> <span class="s1">&#39;bed&#39;</span><span class="p">:</span>
        <span class="n">homer_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">],</span>
                                 <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">],</span>
                                 <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">])[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">],</span>
                                 <span class="s1">&#39;index&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">peaks</span><span class="p">)),</span>
                                 <span class="s1">&#39;useless&#39;</span><span class="p">:</span> <span class="s1">&#39;.&#39;</span><span class="p">,</span>
                                 <span class="s1">&#39;strand&#39;</span><span class="p">:</span><span class="s1">&#39;.&#39;</span><span class="p">})</span>
    
    <span class="n">homer_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">out_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span></div>




<div class="viewcode-block" id="run_HOMER_motif"><a class="viewcode-back" href="../../api.html#treasmo.tl.run_HOMER_motif">[docs]</a><span class="k">def</span> <span class="nf">run_HOMER_motif</span><span class="p">(</span><span class="n">peaks</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">,</span> <span class="n">prefix</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">,</span> 
                    <span class="n">homer_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">split_symbol</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;-&#39;</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="mi">200</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to run Homer from Python script.</span>
<span class="sd">    It will prepare Homer required input file</span>
<span class="sd">    and output results in the directory specified</span>

<span class="sd">    Arguments</span>
<span class="sd">    ---------</span>
<span class="sd">        peaks: List, numpy.array</span>
<span class="sd">            peaks of interests</span>

<span class="sd">        out_dir: str</span>
<span class="sd">            output directory to save the results</span>
<span class="sd">    </span>
<span class="sd">        prefix: str</span>
<span class="sd">            prefix of all files, folder called out_dir/homer_[prefix] will be created to save all the results</span>

<span class="sd">        ref_genome: str</span>
<span class="sd">            reference genome name, e.g., &#39;hg19&#39;, &#39;hg38&#39;</span>

<span class="sd">        homer_path: str</span>
<span class="sd">            path to Homer software, if Homer already added to the PATH, argument can be ignored</span>

<span class="sd">        split_symbol: List[str, str]</span>
<span class="sd">            how peak location ID is formatted</span>
<span class="sd">                &#39;chr1-12345-23456&#39; - split_symbol=[&#39;-&#39;,&#39;-&#39;]</span>
<span class="sd">                &#39;chr1:12345-23456&#39; - split_symbol=[&#39;:&#39;,&#39;-&#39;]</span>
<span class="sd">        </span>
<span class="sd">        size: int</span>
<span class="sd">            Homer paramter, the size of the region used for motif finding</span>


<span class="sd">    Returns</span>
<span class="sd">    ---------</span>
<span class="sd">        DataFrame</span>
<span class="sd">            Homer results saved in out_dir/homer_[prefix]</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">homer_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;index&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">peaks</span><span class="p">)),</span>
                             <span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">],</span>
                             <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">],</span>
                             <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">split_symbol</span><span class="p">[</span><span class="mi">1</span><span class="p">])[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">],</span>
                             <span class="s1">&#39;strand&#39;</span><span class="p">:</span><span class="s1">&#39;.&#39;</span><span class="p">})</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">out_dir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span>

    <span class="n">out_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">.txt&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">prefix</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Save peaks list in BED file format at </span><span class="si">%s</span><span class="s2">&quot;</span><span class="o">%</span><span class="n">out_file</span><span class="p">)</span>
    <span class="n">homer_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">out_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">homer_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="s1">&#39;homer_</span><span class="si">%s</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">prefix</span><span class="p">))</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">homer_dir</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Creat HOMER output folder at </span><span class="si">%s</span><span class="s2">&quot;</span><span class="o">%</span><span class="n">homer_dir</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">homer_dir</span><span class="p">)</span>


    <span class="k">if</span> <span class="n">homer_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;findMotifsGenome.pl </span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> -size </span><span class="si">%s</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">out_file</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">,</span> <span class="n">homer_dir</span><span class="p">,</span> <span class="n">size</span><span class="p">))</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;HOMER run failed. The HOMER peak format DataFrame is returned.&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">homer_df</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">homer_cmd_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homer_path</span><span class="p">,</span><span class="s1">&#39;bin/findMotifsGenome.pl&#39;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> -size </span><span class="si">%s</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">homer_cmd_path</span><span class="p">,</span> <span class="n">out_file</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">,</span> <span class="n">homer_dir</span><span class="p">,</span> <span class="n">size</span><span class="p">))</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;HOMER run failed. The HOMER peak format DataFrame is returned.&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">homer_df</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;HOMER finished successfully! Please check the HTML report for interesting motifs.&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;motif_summary can be run with the motif index for further analysis.&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">homer_df</span></div>




<div class="viewcode-block" id="motif_summary"><a class="viewcode-back" href="../../api.html#treasmo.tl.motif_summary">[docs]</a><span class="k">def</span> <span class="nf">motif_summary</span><span class="p">(</span><span class="n">peak_file</span><span class="p">,</span> <span class="n">homer_dir</span><span class="p">,</span> <span class="n">motif_index</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">,</span>
                  <span class="n">homer_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">200</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to extract related peaks from motifs of interests.</span>


<span class="sd">    Arguments</span>
<span class="sd">    ---------</span>
<span class="sd">        peak_file: str</span>
<span class="sd">            out_dir/prefix.peaks.bed file generated in run_HOMER_motif()</span>

<span class="sd">        homer_dir: str</span>
<span class="sd">            out_dir/homer_prefix in run_HOMER_motif() | Homer output folder</span>
<span class="sd">    </span>
<span class="sd">        motif_index: int</span>
<span class="sd">            motif of interests index in homer_dir/knownResults.html</span>

<span class="sd">        ref_genome: str</span>
<span class="sd">            reference genome name, e.g., &#39;hg19&#39;, &#39;hg38&#39;</span>

<span class="sd">        homer_path: str</span>
<span class="sd">            path to Homer software, if Homer already added to the PATH, argument can be ignored</span>
<span class="sd">        </span>
<span class="sd">        size: int</span>
<span class="sd">            Homer paramter, the size of the region used for motif finding</span>
<span class="sd">            Please keep it the same as in run_HOMER_motif()</span>


<span class="sd">    Returns</span>
<span class="sd">    ---------</span>
<span class="sd">        Motif related peaks information saved in homer_dir/</span>
<span class="sd">        </span>
<span class="sd">        DataFrame</span>
<span class="sd">            contains peak list and motif matching quality information</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">motif_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homer_dir</span><span class="p">,</span> <span class="s1">&#39;knownResults/known</span><span class="si">%s</span><span class="s1">.motif&#39;</span><span class="o">%</span><span class="n">motif_index</span><span class="p">)</span>
    <span class="n">motif_out</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homer_dir</span><span class="p">,</span> <span class="s1">&#39;motif</span><span class="si">%s</span><span class="s1">.peaks.txt&#39;</span><span class="o">%</span><span class="n">motif_index</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">homer_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;findMotifsGenome.pl </span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> -find </span><span class="si">%s</span><span class="s1"> -size </span><span class="si">%s</span><span class="s1"> &gt; </span><span class="si">%s</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span>
                <span class="n">peak_file</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">,</span> <span class="n">homer_dir</span><span class="p">,</span> <span class="n">motif_file</span><span class="p">,</span> <span class="n">size</span><span class="p">,</span> <span class="n">motif_out</span><span class="p">))</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;HOMER run failed. Make sure homer_path is set and all file paths are right.&quot;</span><span class="p">)</span>
            <span class="k">return</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">homer_cmd_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homer_path</span><span class="p">,</span><span class="s1">&#39;bin/findMotifsGenome.pl&#39;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> -find </span><span class="si">%s</span><span class="s1"> -size </span><span class="si">%s</span><span class="s1"> &gt; </span><span class="si">%s</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span>
                <span class="n">homer_cmd_path</span><span class="p">,</span> <span class="n">peak_file</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">,</span> <span class="n">homer_dir</span><span class="p">,</span> <span class="n">motif_file</span><span class="p">,</span> <span class="n">size</span><span class="p">,</span> <span class="n">motif_out</span><span class="p">))</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;HOMER run failed. Make sure homer_path is set and all file paths are right.&quot;</span><span class="p">)</span>
            <span class="k">return</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;HOMER finished successfully! Motif related peaks will be loaded and returned.&quot;</span><span class="p">)</span>
    <span class="n">motif_peaks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">motif_out</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">peak_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">peak_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

    <span class="n">peaks</span> <span class="o">=</span> <span class="n">peak_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">motif_peaks</span><span class="p">[</span><span class="s1">&#39;PositionID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
    <span class="n">peaks</span> <span class="o">=</span> <span class="n">peaks</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">peaks</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">peaks</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">motif_peaks</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;peaks&#39;</span><span class="p">,</span> <span class="n">peaks</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">motif_peaks</span></div>









</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Chaozhong Liu
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2023, Zhandong Liu&#39;s Lab at Baylor College of Medicine.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=365ca57ee442770a23c6"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=365ca57ee442770a23c6"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>